### PR TITLE
Support GraphQL errors spec

### PR DESCRIPTION
**Pull request by Dorokhov**

### CODE REVIEW

File: graphql/executor.go [hunk 1]
```diff
@@ -9,22 +9,57 @@ import (
...
+	Extensions errorExtensions `json:"extensions"`
+}
+
+func newGraphQLError(err error) graphQLError {
+	return newGraphQLErrorRecursive(err, 0)
+}
+
+func newGraphQLErrorRecursive(err error, depth int) graphQLError {
+	if depth >= maxDepth {
+		panic("recursion depth exceeded the limit for GraphQL error")
```
Comment (twalwyn) [inline graphql/executor.go]: Do we have to panic here?

File: graphql/http.go [hunk 2]
```diff
@@ -30,14 +31,15 @@ type httpPostBody struct {
...
-	Errors []string    `json:"errors"`
+	Errors interface{} `json:"errors"`
 }
 
 func (h *httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	writeResponse := func(value interface{}, err error) {
 		response := httpResponse{}
 		if err != nil {
-			response.Errors = []string{err.Error()}
+			log.Println("ERROR: ", err)
```
Comment (twalwyn) [inline graphql/http.go:41]: Is this logline here on purpose?

**Review by twalwyn — COMMENTED** [review]
Is it likely that these changes get accepted upstream? Seems there is some Cloudflare-specific extensions here.

**Comment by twalwyn** [comment]
The commit messages are quite unhelpful.

**Review by twalwyn — APPROVED** [review]
Just reviewed GraphQL errors spec. Seems likely thunder would accept these changes.

File: graphql/executor.go [hunk 3]
```diff
@@ -9,22 +9,57 @@ import (
...
+	"time"
 
 	"github.com/samsarahq/thunder/concurrencylimiter"
 	"github.com/samsarahq/thunder/reactive"
 )
 
+const maxDepth = 100
+
+type errorExtensions struct {
+	Code      string    `json:"code"`
```
Comment (twalwyn) [inline graphql/executor.go]: This should probably be omitempty

**Review by twalwyn — COMMENTED** [review]
Might be good to add some tests?

File: graphql/executor.go (hunk 3 above)
Comment (Dorokhov) [reply graphql/executor.go]: added

File: graphql/executor.go (hunk 1 above)
Comment (Dorokhov) [reply graphql/executor.go]: yeah, panic should be fine, error depth is always = 2, so should never happen, this check is more "theoretical"

File: graphql/http.go (hunk 2 above)
Comment (Dorokhov) [reply graphql/http.go:41]: yes, if err is not a ClientError, but internal failure or bug, the actual message should not be send to the user, but we still need to log it for troubleshooting

File: graphql/executor_test.go [hunk 4]
```diff
@@ -3,12 +3,12 @@ package graphql
 import (
 	"context"
 	"errors"
+	"fmt"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/samsarahq/thunder/internal"
```
Comment (twalwyn) [inline graphql/executor_test.go]: Import ordering changes

File: graphql/executor_test.go [hunk 5]
```diff
@@ -192,6 +192,76 @@ func TestRepeatedFragment(t *testing.T) {
...
+			err:  nil,
+			expected: graphQLError{
+				Message: "Internal server error",
+				Extensions: errorExtensions{
+					Code: "INTERNAL_SERVER_ERROR",
+				},
+			},
+		},
+		{
+			name: "should return internal error when err is nil",
```
Comment (twalwyn) [inline graphql/executor_test.go]: Name of this test not correct?

**Review by twalwyn — APPROVED** [review]

File: graphql/executor_test.go (hunk 5 above)
Comment (Dorokhov) [reply graphql/executor_test.go]: fixed

File: graphql/executor_test.go (hunk 4 above)
Comment (Dorokhov) [reply graphql/executor_test.go]: fixed

**Pull request merged**